src/scopesetbuilder.js

```diff
@@ -0,0 +1,179 @@
+const {scopeCompare, mergeScopeSets, normalizeScopeSet} = require('taskcluster-lib-scopes');
+
+/**
+ * A ScopeMergeTree is an object-structure that holds a tree of scope sets as
+ * they are being merged.
+ *
+ * Merging a large set of normalized scope-sets using mergeScopeSets(A, B) is
+ * expensive as it can only merge two scope-sets at a time, and allocates a new
+ * array for all intermediate results. The complexity of merging N scopes split
+ * between M scope-sets approaches something like O(N * M).
+ *
+ * Instead we shall create a binary tree of nodes where each leaf holds a
+ * scope-set, then we compute the next value for internal node by taking the
+ * value from a child. This leaves us with an average complexity closer
+ * something like O(N * log(M)), notes this just not proven math, merely a
+ * casual guess backed by benchmarks.
+ *
+ * To make it even faster we allow nodes to run a different node (or this) from
+ * the node.next() function, such that the return value is the tree after
+ * consumption of the current value. As a final technicality we leave node.value
+ * undefined until the first call of node.next(), such that we can support empty
+ * nodes like this, by returning null from node.next().
+ */
+class EmptyNode {
+  next() { return null; }
+};
```

Comment:
class appears to be unused..

Reply:
At one point I had the others derive from this one... I could add that back, just because it's pretty :)

The goal was to document the interface that these nodes must implement.